File: patterns/10_proxy/python/legatus.py
#!/usr/bin/env python3
"""
PROXY PATTERN — The Imperial Legatus
"Legatus vocem imperatoris habet" — The delegate has the emperor's voice
Run: python3 legatus.py
"""
from abc import ABC, abstractmethod
from typing import Optional

class IImperialService(ABC):
    @abstractmethod
    def issue_decree(self, province: str, decree: str) -> str: ...
    @abstractmethod
    def get_census(self, province: str) -> str: ...

class EmperorHadrian(IImperialService):
    def __init__(self):
        print("  👑 Emperor Hadrian awakens from his afternoon nap!")
    def issue_decree(self, p: str, d: str) -> str:
        return f"👑 IMPERIAL DECREE for {p}: '{d}' — Signed, Hadrianus P.P."
    def get_census(self, p: str) -> str:
        return f"👑 Census of {p}: Population ~50,000 (all paying taxes... mostly)"

class LegateProxy(IImperialService):
    AUTHORIZED = {"Syria","Aegyptus","Britannia","Gallia","Hispania"}
    def __init__(self):
        self._emperor: Optional[EmperorHadrian] = None
        self._cache: dict[str,str] = {}
        self._count = 0
    def _get_emperor(self) -> EmperorHadrian:
        if self._emperor is None:
            print("  [Proxy] First real request — waking the Emperor...")
            self._emperor = EmperorHadrian()
        return self._emperor
    def _log(self, action: str):
        self._count += 1
        print(f"  [Log #{self._count}] {action}")
    def issue_decree(self, province: str, decree: str) -> str:
        self._log(f"Decree request for '{province}'")
        if province not in self.AUTHORIZED:
            return f"🚫 ACCESS DENIED: '{province}' is not an authorized province!"
        return self._get_emperor().issue_decree(province, decree)
    def get_census(self, province: str) -> str:
        self._log(f"Census request for '{province}'")
        if province in self._cache:
            print(f"  [Cache HIT] Returning cached census for {province}")
            return self._cache[province]
        result = self._get_emperor().get_census(province)
        self._cache[province] = result
        return result
    @property
    def request_count(self) -> int: return self._count

def main():
    print("╔═══════════════════════════════════════════════╗")
    print("║   PROXY PATTERN — Python                      ║")
    print("║   The Imperial Legatus                        ║")
    print("╚═══════════════════════════════════════════════╝\n")
    svc: IImperialService = LegateProxy()
    print("── PROTECTION PROXY: Unauthorized Request ──────")
    print(svc.issue_decree("Barbaria", "Build roads!"))
    print("\n── AUTHORIZED REQUESTS (lazy-init Emperor) ─────")
    print(svc.issue_decree("Britannia", "Build Hadrian's Wall"))
    print(svc.issue_decree("Syria",     "Increase grain quota"))
    print("\n── CACHING PROXY: Census Requests ──────────────")
    print(svc.get_census("Aegyptus"))
    print("\n  (Second call — same province)")
    print(svc.get_census("Aegyptus"))
    print(f"\n  Total requests: {svc.request_count}")
    print("  Emperor woken ONCE (lazy init)")
    print('"Legatus vocem imperatoris habet!"')

if __name__ == "__main__":
    main()

File: patterns/10_proxy/php/legatus.php
<?php
/**
 * PROXY PATTERN — The Imperial Legatus
 * Run: php legatus.php
 */
declare(strict_types=1);

interface IImperialService {
    public function issueDecree(string $province, string $decree): string;
    public function getCensus(string $province): string;
}
class EmperorHadrian implements IImperialService {
    public function __construct() { echo "  👑 Emperor Hadrian awakens from his nap!\n"; }
    public function issueDecree(string $p, string $d): string {
        return "👑 IMPERIAL DECREE for {$p}: '{$d}' — Signed, Hadrianus P.P.";
    }
    public function getCensus(string $p): string {
        return "👑 Census of {$p}: Population ~50,000 (all paying taxes...)";
    }
}
class LegateProxy implements IImperialService {
    private ?EmperorHadrian $emperor = null;
    private array $cache = [];
    private int   $count = 0;
    private array $authorized = ["Syria","Aegyptus","Britannia","Gallia","Hispania"];
    private function getEmperor(): EmperorHadrian {
        if ($this->emperor === null) {
            echo "  [Proxy] First real request — waking the Emperor...\n";
            $this->emperor = new EmperorHadrian();
        }
        return $this->emperor;
    }
    private function log(string $action): void {
        echo "  [Log #" . ++$this->count . "] {$action}\n";
    }
    public function issueDecree(string $p, string $d): string {
        $this->log("Decree for '{$p}'");
        if (!in_array($p, $this->authorized))
            return "🚫 ACCESS DENIED: '{$p}' is not an authorized province!";
        return $this->getEmperor()->issueDecree($p, $d);
    }
    public function getCensus(string $p): string {
        $this->log("Census for '{$p}'");
        if (isset($this->cache[$p])) {
            echo "  [Cache HIT] Returning cached census for {$p}\n";
            return $this->cache[$p];
        }
        $r = $this->getEmperor()->getCensus($p);
        $this->cache[$p] = $r;
        return $r;
    }
    public function getCount(): int { return $this->count; }
}

echo "╔═══════════════════════════════════════════════╗\n";
echo "║   PROXY PATTERN — PHP                         ║\n";
echo "║   The Imperial Legatus                        ║\n";
echo "╚═══════════════════════════════════════════════╝\n\n";
$svc = new LegateProxy();
echo "── PROTECTION: Unauthorized ────────────────────\n";
echo $svc->issueDecree("Barbaria", "Build roads!") . "\n";
echo "\n── AUTHORIZED REQUESTS ─────────────────────────\n";
echo $svc->issueDecree("Britannia", "Build Hadrian's Wall") . "\n";
echo "\n── CACHING: Census ──────────────────────────────\n";
echo $svc->getCensus("Aegyptus") . "\n";
echo "\n  (Second call)\n";
echo $svc->getCensus("Aegyptus") . "\n";
echo "\n  Total requests: " . $svc->getCount() . "\n";
echo "\"Legatus vocem imperatoris habet!\"\n";
